### PR TITLE
Add ThreadJob module package and tests

### PR DESCRIPTION
**Pull request by PaulHigin**
## PR Summary

This adds the ThreadJob module package from the PSGallery, along with Pester tests

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by TravisEz13 — APPROVED** [review]

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (anmenaga) [inline test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: This is not related to review; but for my education:
The ThrottleLimit syntax looks interesting; couple of questions:
* What is the scope of ThrottleLimit ? e.g. "all threadjobs in current script" ?
* How do subsequent Start-ThreadJob's in the script know under what ThrottleLimit they operate?

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (anmenaga) [inline test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: I would rename the Describe to "Job2 class API tests" or something like this.
For an unfamiliar person it is not clear what 'Job2' is.

**Review by anmenaga — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Can we use Pester 4 syntax? 
```PowerShell
$results | Should -Be "Hello"
```

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Can we also add a test for a non-terminating error where some objects are returned and 1 non-terminating error is returned?

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Is this test same as https://github.com/PowerShell/PowerShell/pull/7169/files#diff-74d73c8bffa8cd97ca36a864c1dc71bbR172

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: extra empty line

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: extra empty line

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: extra empty line

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: This is not intended to be an exhaustive list of job tests, but I have added data stream tests.

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Throttle limit is per PowerShell session so it limits all concurrent threadjobs created in that session.  The queue is static (singleton).  I currently don't have a way to get the current throttle limit value ... you just set it.  But something like that can be added.

See source code for more details:
https://github.com/PaulHigin/PSThreadJob/blob/master/ThreadJob/ThreadJob/PSThreadJob.cs#L18

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Ok.

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: This is outside the scope of this change.  Please create an Issue for adding more threadjob tests.

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: The build only pulls in Pester 3.3.9 right now which does not support this syntax.  We should wait to update tests until the default pester version supports the new syntax.

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: It is a slightly different test since it involves Receive-Job, so I'd like to keep it.

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Thanks, but will keep for readability.

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Thanks, but will keep for readability.

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Thanks, but will keep for readability.

**Review by anmenaga — APPROVED** [review]

**Comment by TravisEz13** [comment]
@adityapatwardhan can you update your review?

**Review by BrucePay — APPROVED** [review]

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Good point.  It just takes a scriptblock.  But I used the same parameter name as what Start-Job uses, which also just takes a scriptblock.  So for consistency we should probably keep the parameter name as is (or change both).

Comment (BrucePay) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: I was just thinking of consistency between `-ScriptBlock` and `-InitialScriptBlock`. But following what existing commands do is probably fine.

**Comment by PaulHigin** [comment]
@TravisEz13  I just found a bug in this version of ThreadJob.  Please hold off merging until I have a fix ready.  Thanks.

**Comment by TravisEz13** [comment]
@PaulHigin  to prevent maintainers from merging, rename your PR with WIP in front of the title.  remove WIP to unblock.  This will fail the Work In Progress (WIP), CI test immediately.

**Comment by PaulHigin** [comment]
I have fixed the bug, updated tests, and updated the package with the new version.  Please take a look.

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by PaulHigin** [comment]
@adityapatwardhan  Can you please review changes?

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: This should be `Should -BeNullOrEmpty`

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1
Comment (PaulHigin) [reply test/powershell/Modules/ThreadJob/ThreadJob.Tests.ps1]: Fixed

**Review by adityapatwardhan — APPROVED** [review]

File: assets/files.wxs
Comment (TravisEz13) [inline assets/files.wxs]: Not an issue for this PR.  We should consider not putting the modules in a version folder.  If we have to update the modules, this will make patching very difficult.

Comment (TravisEz13) [reply assets/files.wxs]: Filed https://github.com/PowerShell/PowerShell/issues/7236

**Review by TravisEz13 — APPROVED** [review]

**Pull request merged**
